### PR TITLE
2122 querying root dataverse contents (and other permission performance boosts)

### PR DESCRIPTION
**Pull request by oscardssmith**
This speeds up querying root dataverse contents by adding fast paths for non users and superusers who now don't get permissions checked, as well as speedifying `permissionServiceBean` with shortcut has methods that are more efficient than the methods to find all permissions for a user. I also removed redundant recursive calls to `groupsFor` which speeds things up.

These changes make the query have acceptable performance for non-users and superusers, and speeds up the time of a regular user from ????????? to 26 minutes (on my laptop). This level of speedup is probably also achieved for most other calls to `permissionServiceBean`, which may significantly speed up other unrelated pieces of the code.

The main thing this PR still needs is a set of integration tests for permissions involving groups and nested groups with inherited permissions to make sure there are no functional regressions. 

In addition caching of recent `groupsFor` calls should be able to speed up the remaining case of users who are not superusers to a reasonable level.

- connects to #2122:querying root dataverse contents

- [ ] Integration tests:

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: https://github.com/IQSS/dataverse/tree/develop/scripts/database/upgrades
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java [hunk 1]
```diff
@@ -184,96 +199,168 @@ public boolean canIssueCommand(String commandName) throws ClassNotFoundException
...
         }
 
     }
 
+    public Collection<DvObject> userPermissions(User user, DvObject dvo){
+        String powerfull_roles = "select id from dataverserole where (permissionbits&12)!=0";
+        List<Integer> roles = em.createNativeQuery(powerfull_roles).getResultList();
+        String x = "select id from dataverserole where (permissionbits&12)!=0";
+        return null;
+    }
```
Comment (oscardssmith) [inline src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java]: note to self: delete this method

File: src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java [hunk 2]
```diff
@@ -184,96 +199,193 @@ public boolean canIssueCommand(String commandName) throws ClassNotFoundException
...
+
+        return false;
+    }
+
+    private boolean hasPermissionsForSingleRoleAssignee(RoleAssignee ra, DvObject d, Set<Permission> p) {
+        // File special case.
+        if (d instanceof DataFile && p.contains(Permission.DownloadFile)) {
+            // unrestricted files that are part of a release dataset 
+            // automatically get download permission for everybody:
+            //      -- L.A. 4.0 beta12
```
[the remark this answers is not in the document]
Comment (oscardssmith) [reply src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java]: don't think they're stale

File: src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java [hunk 3]
```diff
@@ -184,96 +199,193 @@ public boolean canIssueCommand(String commandName) throws ClassNotFoundException
...
+    public Set<Permission> permissionsFor(DataverseRequest req, DvObject dvo) {
+        // Start with permissions specifically given to the user
+        Set<Permission> permissions = permissionsForSingleRoleAssignee(req.getUser(), dvo);
+
         // Add permissions gained from groups
-        for ( Group g : groups ) {
-            final Set<Permission> groupPremissions = permissionsForSingleRoleAssignee(g,dvo);
-            permissions.addAll(groupPremissions);
+        for (Group g : groupService.groupsFor(req, dvo)) {
+            permissionsForSingleRoleAssignee(g, dvo, permissions);
```
[the remark this answers is not in the document]
Comment (oscardssmith) [reply src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java]: yeah, I just realized I introduced 2 different `permissionForSingleRoleAssignee` that pass sets with completely different meanings. Will fix.

File: src/main/java/edu/harvard/iq/dataverse/util/TimeoutCache.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+import java.util.AbstractMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Logger;
+
+public class TimeoutCache<K, V> implements Map<K, V> {
```
[the remark this answers is not in the document]
Comment (oscardssmith) [reply src/main/java/edu/harvard/iq/dataverse/util/TimeoutCache.java]: I figured that could wait until we made a final decision to use it.

File: src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthew-a-dunlap) [reply src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java]: Probably just take out the L.A. 4.0 tag or rewrite it... makes it seem like the code is ancient yet its new.

File: src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java [hunk 5]
```diff
@@ -50,24 +50,24 @@
...
     private static final Logger logger = Logger.getLogger(PermissionServiceBean.class.getName());
-    
-    private static final Set<Permission> PERMISSIONS_FOR_AUTHENTICATED_USERS_ONLY = 
-           EnumSet.copyOf(Arrays.asList(Permission.values()).stream()
-                    .filter( Permission::requiresAuthenticatedUser )
-                    .collect( Collectors.toList() ));
-    
+
+    private static final Set<Permission> PERMISSIONS_FOR_AUTHENTICATED_USERS_ONLY
+            = EnumSet.copyOf(Arrays.asList(Permission.values()).stream()
```
Comment (benjamin-martinez) [inline src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java]: I'm assuming this was done on purpose/is good syntax.

File: src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java [hunk 6]
```diff
@@ -218,54 +217,64 @@ public boolean has(String pName) {
...
+     * @param required (sub)set of permissions {@code req} has on the objects in the returned list
+     * @return list of {@code dvo} children over which {@code req} has at least {@code required} permissions.
+     */
     public List<DvObject> whichChildrenHasPermissionsFor(DataverseRequest req, DvObjectContainer dvo, Set<Permission> required) {
         List<DvObject> children = dvObjectServiceBean.findByOwnerId(dvo.getId());
-        
         User user = req.getUser();
+        
+        // quick cases
         if (user.isSuperuser()) {
```
Comment (oscardssmith) [inline src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java]: should a required isempty also go here?

**Comment by pameyer** [comment]
on ac11957f37e3615a851020f9762bb97f75ed9842:
``
Failed tests:   testDeleteDatasetWhileFileIngesting(edu.harvard.iq.dataverse.api.DatasetsIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testMoveDataverse(edu.harvard.iq.dataverse.api.DataversesIT): Expected status code <200> doesn't match actual status code <500>.(..)
  testJsonParserWithDirectoryLabels(edu.harvard.iq.dataverse.api.FileMetadataIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testJsonParserWithDirectoryLabels(edu.harvard.iq.dataverse.api.FileMetadataIT): Expected status code <200> doesn't match actual status code <404>.(..)
  test_001_AddFileGood(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testAddFileBadJson(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  test_005_AddFileBadPermissions(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  test_006_ReplaceFileGood(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  test_006_ReplaceFileGoodTabular(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testForceReplace(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  test_007_ReplaceFileUnpublishedAndBadIds(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  test_008_ReplaceFileAlreadyDeleted(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testReplaceFileBadJson(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testAddTinyFile(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testRestrictFile(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testRestrictAddedFile(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testAccessFacet(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testCuratorSendsCommentsToAuthor(edu.harvard.iq.dataverse.api.InReviewWorkflowIT): XML path message doesn't match.(..)
  testSearchPermisions(edu.harvard.iq.dataverse.api.SearchIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testCreateDataverseCreateDatasetUploadFileDownloadFileEditTitle(edu.harvard.iq.dataverse.api.SwordIT): Expected status code <400> doesn't match actual status code <500>.(..)
  testCreateAndDeleteDatasetInRoot(edu.harvard.iq.dataverse.api.SwordIT): Expected status code <201> doesn't match actual status code <500>.(..)
  testCreateDatasetPublishDestroy(edu.harvard.iq.dataverse.api.SwordIT): Expected status code <201> doesn't match actual status code <500>.(..)

Tests in error: 
  testCreateDataset(edu.harvard.iq.dataverse.api.DatasetsIT): Failed to parse the JSON document
  testAddUpdateDatasetViaNativeAPI(edu.harvard.iq.dataverse.api.DatasetsIT): Failed to parse the JSON document
  testSequentialNumberAsIdentifierGenerationStyle(edu.harvard.iq.dataverse.api.DatasetsIT): Failed to parse the JSON document
  testPrivateUrl(edu.harvard.iq.dataverse.api.DatasetsIT): Failed to parse the JSON document
  testFileChecksum(edu.harvard.iq.dataverse.api.DatasetsIT): Failed to parse the JSON document
  testSearchCitation(edu.harvard.iq.dataverse.api.SearchIT): Failed to parse the JSON document
  testDatasetThumbnail(edu.harvard.iq.dataverse.api.SearchIT): Failed to parse the JSON document
  testIdentifier(edu.harvard.iq.dataverse.api.SearchIT): Failed to parse the JSON document
  testDeleteFiles(edu.harvard.iq.dataverse.api.SwordIT): Failed to parse the XML document

Tests run: 74, Failures: 22, Errors: 9, Skipped: 0
``

**Comment by oscardssmith** [comment]
Honestly, ~90% of the changes here could be removed while maintaining the performance improvement. I would do it but i no longer have permissions.

**Comment by landreev** [comment]
Thanks @pameyer, for the detailed report. 
I checked in some fixes for all of the test failures below - EXCEPT ONE. Will explain in detail in the issue (#2122)

> on [ac11957](https://github.com/IQSS/dataverse/commit/ac11957f37e3615a851020f9762bb97f75ed9842):
> ``
> Failed tests: testDeleteDatasetWhileFileIngesting(edu.harvard.iq.dataverse.api.DatasetsIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testMoveDataverse(edu.harvard.iq.dataverse.api.DataversesIT): Expected status code <200> doesn't match actual status code <500>.(..)
> testJsonParserWithDirectoryLabels(edu.harvard.iq.dataverse.api.FileMetadataIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testJsonParserWithDirectoryLabels(edu.harvard.iq.dataverse.api.FileMetadataIT): Expected status code <200> doesn't match actual status code <404>.(..)
> test_001_AddFileGood(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testAddFileBadJson(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> test_005_AddFileBadPermissions(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> test_006_ReplaceFileGood(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> test_006_ReplaceFileGoodTabular(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testForceReplace(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> test_007_ReplaceFileUnpublishedAndBadIds(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> test_008_ReplaceFileAlreadyDeleted(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testReplaceFileBadJson(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testAddTinyFile(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testRestrictFile(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testRestrictAddedFile(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testAccessFacet(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testCuratorSendsCommentsToAuthor(edu.harvard.iq.dataverse.api.InReviewWorkflowIT): XML path message doesn't match.(..)
> testSearchPermisions(edu.harvard.iq.dataverse.api.SearchIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testCreateDataverseCreateDatasetUploadFileDownloadFileEditTitle(edu.harvard.iq.dataverse.api.SwordIT): Expected status code <400> doesn't match actual status code <500>.(..)
> testCreateAndDeleteDatasetInRoot(edu.harvard.iq.dataverse.api.SwordIT): Expected status code <201> doesn't match actual status code <500>.(..)
> testCreateDatasetPublishDestroy(edu.harvard.iq.dataverse.api.SwordIT): Expected status code <201> doesn't match actual status code <500>.(..)
> 
> Tests in error:
> testCreateDataset(edu.harvard.iq.dataverse.api.DatasetsIT): Failed to parse the JSON document
> testAddUpdateDatasetViaNativeAPI(edu.harvard.iq.dataverse.api.DatasetsIT): Failed to parse the JSON document
> testSequentialNumberAsIdentifierGenerationStyle(edu.harvard.iq.dataverse.api.DatasetsIT): Failed to parse the JSON document
> testPrivateUrl(edu.harvard.iq.dataverse.api.DatasetsIT): Failed to parse the JSON document
> testFileChecksum(edu.harvard.iq.dataverse.api.DatasetsIT): Failed to parse the JSON document
> testSearchCitation(edu.harvard.iq.dataverse.api.SearchIT): Failed to parse the JSON document
> testDatasetThumbnail(edu.harvard.iq.dataverse.api.SearchIT): Failed to parse the JSON document
> testIdentifier(edu.harvard.iq.dataverse.api.SearchIT): Failed to parse the JSON document
> testDeleteFiles(edu.harvard.iq.dataverse.api.SwordIT): Failed to parse the XML document
> 
> Tests run: 74, Failures: 22, Errors: 9, Skipped: 0
> ``

**Comment by landreev** [comment]
@oscardssmith  
Hi Oscar, 
Thanks for the offer of cleanup. 
When you say that "90% of the changes here could be removed" - was that the plan for what Michael was going to do with this branch? So Michael didn't have time to work on it, so the PR was sitting there with his name on it; and I felt increasingly bad about not being able to merge it, so I finally picked it up. 
But note that the cleanup was not the most pressing issue with this branch; it was put on hold because of all the failing tests (that Pete reported in early Sep., above). 
I spent some time looking into it; most of the errors and 500s were caused by just 2 issues with 2 named queries. 
Now all of the tests from the list are passing, except for this one: testCreateDatasetPublishDestroy in SwordIT - and this one appears to be a real problem. 
It's no longer throwing a 500, as in the original report (that was fixed by one of my fixes). It goes further and bombs on line 594, when the test tries to list the contents of a dataverse that it has created, populated and published. So the error is right in the API for which this issue was opened: 
/api/dataverses/.../contents
And it actually appears to be a permission issue. 
Once again, it's a published dataverse, with a published dataset, so everybody should be able to see it. 
If I list its contents as the superuser, like this: 
```
 curl "http://localhost:8080/api/dataverses/dv56233a79/contents?key=SUPERUSER_KEY"
```
I get back its contents correctly. However, if I try it as the user who created the dataverse and the dataset (and this is what's happening in the test that is failing), I get nothing:
```
curl "http://localhost:8080/api/dataverses/dv56233a79/contents?key=USER_WHO_CREATED_THE_DATAVERSE"
{"status":"OK","data":[]}
```
(same thing when I try it without any api key). 
Once again, this is the API for which the issue was originally opened, and it doesn't appear to be working. 
The entries in the RoleAssignment table appear to be correct, for the dataverse and dataset in question. But both are published - so should be visible to everybody anyway. 

I can't immediately tell what's going on. I was hoping it was some kind of an obvious conflict with something that's been introduced into the branch since you worked on it, as it was synced up with develop; but I'm not seeing anything obvious. 

Do you have any pointers? 
If we have already taken you off the dev group on github, I can definitely add you back. To be clear, no pressure/expectations for you to actually work on this (obviously). But if you have anything to suggest or contribute, it'll be appreciated. 

I'm sure I can figure it out. But still hoping it may be something simple you could point out immediately...

Hope you're well! 

> Honestly, ~90% of the changes here could be removed while maintaining the performance improvement. I would do it but i no longer have permissions.

**Comment by pdurbin** [comment]
@landreev just checking that you know about pull request #4998 which includes commits from @oscardssmith (which are also in this pull request #4883) followed by commits by @michbarsinai

**Comment by landreev** [comment]
Oh, yes, what is the relationship between this branch and #4998, does it have everything this branch has? Should we abandon this branch? 
I'll test the functionality in #4998 and see if the API in question is working there; (but my understanding was that that PR was for cleanup only; and should not affect how it actually works... -?)

**Comment by oscardssmith** [comment]
What I would recommend is abandoning both this and michael's PR's. Then cherry pick the docs changes and the change to how group transitive closure is computed (including removing the double recursion), and leaving the rest. Those changes account for almost all of the functional improvements.

**Comment by oscardssmith** [comment]
The rewritten method for child permissions might be nice too as that gives a notable speedup for that 1 case.

**Comment by michbarsinai** [comment]
The branches are identical up to commit  05b704b  (August 10). I #4998 still has those breaking @NotNull annotations, which may cause the "more broken state".

-- Michael

**Comment by landreev** [comment]
@michbarsinai For the record, no this appears to have nothing to do with those NotNull annotations. Adding that fix to the "cleanup" branch doesn't change the behavior in question - you still can't create the initial, root dataverse w/ this build. I'm assuming it's because somewhere in the code something expects the root dataverse to always be present in order to perform any permission checks? - I'm back to debugging the original, "uncleaned-up" branch; but will take another look at this too.

**Comment by pdurbin** [comment]
Now that this pull request has been merged, we can't run the API tests on the phoenix server: #5393

**Comment by kcondon** [comment]
Looks like this is failing, using the builtin secret key. Maybe that was overlooked?
curl -H "Content-type:application/json" -d @data/role-admin.json http://localhost:8080/api/admin/roles/

**Comment by landreev** [comment]
@oscardssmith 
Yes, I'm glad we were able to move this PR finally; thanks for all you work on it again. 
There is unfortunately another problem we have somehow introduced (above); with tests failing on the dev. branch since this was merged. The entire restassured test set and the setup scripts were definitely working for me 2 weeks ago, when I was actively working on the branch; so must be a conflict introduced in one of the most recent syncs up with the develop branch... Will be looking into it today. 

But, answering your question, yes, I ran some performance tests; I made a  script for recursive crawling, and ran it on the entire prod. dataverse, as different users, and on individual dataverses. I'll post the numbers; they seem sensible, but most importantly it is something you can do now - where it was just not practical at all to crawl the entire root dataverse holdings. 

(I'll post the number in the issue, not here)

I haven't tested the effects on reindexing; I probably should, and will. But I honestly don't expect that much of an effect there. Full reindexing is always done as the superuser; and there is no recursive crawling involved (instead of going through dataverses recursively, it just gets a linear list of all the datasets from the database, and goes through them). I also believe we have some evidence that shows that most of the time there is actually spent doing the indexing... But as I said, I'll test it, to know for sure.

**Comment by landreev** [comment]
@pdurbin et al - there's a chance the last night's test failure was a result of a jenkins and/or phoenix fluke; I've added a comment in #5393.

**Pull request merged**